design/src/rfcs/rfc0023_serialization_and_deserialization.md

```diff
@@ -0,0 +1,90 @@
+<!-- Give your RFC a descriptive name saying what it would accomplish or what feature it defines -->
+RFC: Serialization and Deserialization
+=============
+
+<!-- RFCs start with the "RFC" status and are then either "Implemented" or "Rejected".  -->
+> Status: RFC
+>
+> Applies to: builder structs and other data types that are connected to builder structs
+
+# Overview
+There has been some github issues requesting for the implementation of Serialize/Deserialize traits provided serde crate.  
+At the current state of sdk, user must implement their own solution in order to serialize/deserialize sdk's datatypes.  
+This is a RFC for implementing serde traits.
+
+# Potential Use Cases
+- Testing
+- Reusable conifiguration, infrastructure as a service
+- Logging
+ 
+... and more!
+
+# Implementation
+Add `serde::Serialize` and `serde::Deserialize` to builder datatypes.
+
+# Discussion
+## Enum Representation 
+serde allows programmers to use one of four different tagging ([internal, external, adjacent and untagged](https://serde.rs/enum-representations.html)) when serializing enum.  
+
+I suggest going with `untagged` variant.  
```

Comment:
If we go with `untagged` it might make it hard (or impossible) to differentiate between models that are very similar. Given that, it would be nice to see an example of what serialization and deserialization of something like a DynamoDB [PutItem](https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/client/fluent_builders/struct.PutItem.html) operation would look like from a future user's perspective.

Reply:
_(I don't mean to veto this, I'm just curious how it would work.)_

Reply:
I think I completely messed up here. 

Initially, I thought about using adjacent tagging, but on the last minute before submitting this RFC I changed it to untagged because I mistakenly believed that every enums that you can find in the SDKs looked something like this. 
```rust
// https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/account/src/model.rs
pub enum AlternateContactType {
    #[allow(missing_docs)] // documentation missing in model
    Billing,
    #[allow(missing_docs)] // documentation missing in model
    Operations,
    #[allow(missing_docs)] // documentation missing in model
    Security,
    /// Unknown contains new variants that have been added since this code was generated.
    Unknown(String),
}
```

Reply:
I think I want to replace this section with something like this.

# Discussion
## Enum Representation 
serde allows programmers to use one of four different tagging ([internal, external, adjacent and untagged](https://serde.rs/enum-representations.html)) when serializing enum.  

- untagged  
  You cannot deserialize serialized data in some cases. 
  For example, [aws_sdk_dynamodb::model::AttributeValue](https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/model/enum.AttributeValue.html) has `Null(bool)` and `Bool(bool)`, which you cannot distinguish serialized values without a tag.
  
- internal  
  This results in compile time error.
  [*Using a #[serde(tag = "...")] attribute on an enum containing a tuple variant is an error at compile time*](https://serde.rs/enum-representations.html).

Reply:
Only question that remains is whether you want to use external tagging or adjacent tagging.  

Both of them will work.  

I think we should go with external tagging, since this is the default way and there is less question to answer.

If you want to go with adjacent tagging, you need to figure out what the name of the field is going to be.

Here is an example from the serde's document; You need to tell serde what the name of the field for tag/content is going to be.
```rust
#[derive(Serialize, Deserialize)]
#[serde(tag = "t", content = "c")]
enum Block {
    Para(Vec<Inline>),
    Str(String),
}
```